realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTests.java

```diff
@@ -996,4 +999,187 @@ public void realmListRemoveAllFromRealm() {
         assertEquals(0, list.size());
         assertEquals(0, realm.where(Dog.CLASS_NAME).count());
     }
+
+    // Test if waitForChange gets waked up by 1) empty transaction 2) transaction with some data changes.
+    @Test
+    public void waitForChange() throws InterruptedException {
+        final long numberOfBlocks = 16;
+        final long numberOfObjects = 1; // must be greater than 1
+        final CountDownLatch openedLatch = new CountDownLatch(1);
+        final CountDownLatch changedLatch = new CountDownLatch(1);
+        final CountDownLatch closedLatch = new CountDownLatch(1);
+        final AtomicBoolean result1 = new AtomicBoolean(false);
+        final AtomicBoolean result2 = new AtomicBoolean(false);
+        final AtomicLong count1 = new AtomicLong(0);
+        final AtomicLong count2 = new AtomicLong(0);
+
+        // wait in background
+        final CountDownLatch signalTestFinished = new CountDownLatch(1);
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                DynamicRealm realm = DynamicRealm.getInstance(defaultConfig);
+                openedLatch.countDown();
+                result1.set(realm.waitForChange());
+                count1.set(realm.where(AllTypes.CLASS_NAME).count());
+                changedLatch.countDown();
+                result2.set(realm.waitForChange());
+                count2.set(realm.where(AnnotationIndexTypes.CLASS_NAME).count());
+                realm.close();
+                closedLatch.countDown();
+            }
+        });
+        thread.start();
+
+        TestHelper.awaitOrFail(openedLatch);
+        Thread.sleep(100);
+        realm.beginTransaction();
+        realm.commitTransaction();
+        TestHelper.awaitOrFail(changedLatch);
+        assertTrue(result1.get());
+        assertEquals(0, count1.get());
+        Thread.sleep(500);
+        populateForDistinct(realm, numberOfBlocks, numberOfObjects, false);
+        TestHelper.awaitOrFail(closedLatch);
+        assertTrue(result2.get());
+        assertEquals(numberOfBlocks * numberOfBlocks, count2.get());
+    }
+
+    // Test if waitForChange gets waked up by stopWaitForChange called.
+    @Test
+    public void stopWaitForChange() throws InterruptedException {
+        final CountDownLatch openedLatch = new CountDownLatch(1);
+        final CountDownLatch canceledLatch = new CountDownLatch(1);
+        final CountDownLatch closedLatch = new CountDownLatch(1);
+        final AtomicBoolean result = new AtomicBoolean(false);
+        final DynamicRealm[] otherRealm = new DynamicRealm[1];
+
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                otherRealm[0] = DynamicRealm.getInstance(defaultConfig);
+                openedLatch.countDown();
+                result.set(otherRealm[0].waitForChange());
+                canceledLatch.countDown();
```

Comment:
Hmm, looks like you are right, it is partly unavoidable in this case. I would do this instead though:

```
   @Test
    public void stopWaitForChange() throws InterruptedException {
        final CountDownLatch bgRealmOpened = new CountDownLatch(1);
        final CountDownLatch bgRealmClosed = new CountDownLatch(1);
        final AtomicBoolean waitForChangeResult = new AtomicBoolean(false);
        final AtomicReference<DynamicRealm> bgRealmRef = new AtomicReference(null);

        new Thread(new Runnable() {
            @Override
            public void run() {
                Realm realm = DynamicRealm.getInstance(defaultConfig)
                bgRealmRef.set(realm);
                bgRealmOpened.countDown();
                waitForChangeResult.set(realm.waitForChange());
                realm.close();
                bgRealmClosed.countDown();
            }
        }).start();

        TestHelper.awaitOrFail(bgRealmOpened);
        // Looks like it is unavoidable :( Perhaps see if http://stackoverflow.com/questions/1825623/java-how-to-know-when-thread-is-waiting can help, although I doubt it since the block is native code
        Thread.sleep(200); 
        bgRealmRef.get().stopWaitForChange();
        TestHelper.awaitOrFail(bgRealmClosed);
        assertFalse(result.get());
    }
```